pynamodb/tests/integration/test_transaction_integration.py

```diff
@@ -0,0 +1,252 @@
+    amount = NumberAttribute()
+    currency = UnicodeAttribute()
+
+
+class DifferentRegion(Model):
+
+    class Meta:
+        region = 'us-east-2'
+        table_name = 'different-region'
+
+    entry_index = NumberAttribute(hash_key=True)
+
+
+TEST_MODELS = [
+    BankStatement,
+    DifferentRegion,
+    LineItem,
+    User,
+]
+
+
+@pytest.fixture(scope='module')
+def connection(ddb_url):
+    yield Connection(host=ddb_url)
+
+
+@pytest.fixture(scope='module', autouse=True)
+def create_tables(ddb_url):
+    for m in TEST_MODELS:
+        m.Meta.host = ddb_url
+        m.create_table(
+            read_capacity_units=10,
+            write_capacity_units=10,
+            wait=True
+        )
+
+    yield
+
+    for m in TEST_MODELS:
+        if m.exists():
+            m.delete_table()
+
+
+def get_error_code(error):
+    return error.cause.response['Error'].get('Code')
+
+
+@pytest.mark.ddblocal
+def test_transact_write__error__idempotent_parameter_mismatch(connection):
+    client_token = str(uuid.uuid4())
+
+    with TransactWrite(connection=connection, client_request_token=client_token) as transaction:
+        transaction.save(User(1))
+        transaction.save(User(2))
+
+    with pytest.raises(PutError) as exc_info:
+        # committing the first time, then adding more info and committing again
+        with TransactWrite(connection=connection, client_request_token=client_token) as transaction:
+            transaction.save(User(3))
+    assert get_error_code(exc_info.value) == IDEMPOTENT_PARAMETER_MISMATCH
+
+    # ensure that the first request succeeded in creating new users
+    assert User.get(1)
+    assert User.get(2)
+
+    with pytest.raises(DoesNotExist):
+        # ensure it did not create the user from second request
+        User.get(3)
+
+
+@pytest.mark.ddblocal
+def test_transact_write__error__different_regions(connection):
+    with pytest.raises(PutError) as exc_info:
+        with TransactWrite(connection=connection) as transact_write:
+            # creating a model in a table outside the region everyone else operates in
+            transact_write.save(DifferentRegion(entry_index=0))
+            transact_write.save(BankStatement(1))
+            transact_write.save(User(1))
+    assert get_error_code(exc_info.value) == RESOURCE_NOT_FOUND
+
+
+@pytest.mark.ddblocal
+def test_transact_write__error__transaction_cancelled(connection):
+    # create a users and a bank statements for them
+    User(1).save()
+    BankStatement(1).save()
+
+    # attempt to do this as a transaction with the condition that they don't already exist
+    with pytest.raises(PutError) as exc_info:
+        with TransactWrite(connection=connection) as transaction:
+            transaction.save(User(1), condition=(User.user_id.does_not_exist()))
+            transaction.save(BankStatement(1), condition=(BankStatement.user_id.does_not_exist()))
+    assert get_error_code(exc_info.value) == TRANSACTION_CANCELLED
```

Comment:
Can we check that the sub-code was `ConditionalCheckFailed` here too? I'd also be interested in seeing another test for the `TransactionConflict` case

Reply:
so the error returned by amazon is kinda dumb, and there are [other people](https://github.com/aws/aws-sdk-js/issues/2464) who have worked around it by parsing the string. should we do that to make it more helpful?

an example is, if one condition fails out of 3, the error message is:
```python
{
    'Error': {
        'Message': 'Transaction cancelled, please refer cancellation reasons for specific reasons [None, ConditionalCheckFailed, None]',
        'Code': 'TransactionCanceledException'
    }
}
```

Since the list of errors is kept in request order, we could theoretically return the identifier map with the parsed out error and store it on the response

Reply:
Ah I see, that is unfortunate! I think it's probably OK for the test to just do a `.contains` match to see that it failed for the expected reason, but I don't feel too strongly so whichever way you'd prefer

Reply:
I think I'll hold off on adding the reasons to the response, and create a separate pr for that :0 for now, i can just add a check that the message contains the expected reason within the string